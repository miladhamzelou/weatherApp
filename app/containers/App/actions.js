import {
  LOAD_FORECAST,
  LOAD_FORECAST_SUCCESS,
  LOAD_FORECAST_ERROR,
} from './constants';

/**
 * Load the forecasts, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_FORECAST
 */
export function loadForecast() {
  return {
    type: LOAD_FORECAST,
  };
}

/**
 * Dispatched when the forecasts are loaded by the request saga
 *
 * @param  {object} forecast The repository data
 *
 * @return {object}      An action object with a type of LOAD_FORECAST_SUCCESS passing the forecasts
 */
export function forecastLoaded(forecasts) {
  return {
    type: LOAD_FORECAST_SUCCESS,
    forecasts,
  };
}

/**
 * Dispatched when loading the forecasts fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of  passing the error
 */
export function forecastLoadingError(error) {
  return {
    type: LOAD_FORECAST_ERROR,
    error,
  };
}
