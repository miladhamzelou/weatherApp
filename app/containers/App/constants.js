export const DEFAULT_LOCALE = 'en';
export const LOAD_FORECAST = 'weatherApp/App/LOAD_FORECAST';
export const LOAD_FORECAST_SUCCESS = 'weatherApp/App/LOAD_FORECAST_SUCCESS';
export const LOAD_FORECAST_ERROR = 'weatherApp/App/LOAD_FORECAST_ERROR';

