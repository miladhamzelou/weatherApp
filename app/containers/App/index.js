/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React from 'react';
import Helmet from 'react-helmet';
import styled from 'styled-components';
import withProgressBar from 'components/ProgressBar';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
  background: #ccc;
  height: 80vh;
`;

export function App(props) {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - WeatherApp"
        defaultTitle="WeatherApp"
        meta={[
          { name: 'description', content: 'A simple WeatherApp' },
        ]}
      />
      <div>
        {React.Children.toArray(props.children)}
      </div>
    </AppWrapper>
  );
}

App.propTypes = {
  children: React.PropTypes.node,
};

export default withProgressBar(App);
