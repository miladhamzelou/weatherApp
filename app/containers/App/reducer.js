import { fromJS } from 'immutable';

import {
  LOAD_FORECAST_SUCCESS,
  LOAD_FORECAST,
  LOAD_FORECAST_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentUser: false,
  userData: {
    forecasts: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_FORECAST:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['userData', 'forecasts'], false);
    case LOAD_FORECAST_SUCCESS:
      return state
        .setIn(['userData', 'forecasts'], action.forecasts)
        .set('loading', false);
    case LOAD_FORECAST_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
