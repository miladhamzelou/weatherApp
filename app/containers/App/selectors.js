/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = (state) => state.get('global');

const makeSelectLoading = () => createSelector(
  selectGlobal,
  (globalState) => globalState.get('loading')
);

const makeSelectError = () => createSelector(
  selectGlobal,
  (globalState) => globalState.get('error')
);

const makeSelectForecasts = () => createSelector(
  selectGlobal,
  (globalState) => globalState.getIn(['userData', 'forecasts'])
);

const makeSelectLocationState = () => {
  let prevRoutingState;
  let prevRoutingStateJS;

  return (state) => {
    const routingState = state.get('route'); // or state.route

    if (!routingState.equals(prevRoutingState)) {
      prevRoutingState = routingState;
      prevRoutingStateJS = routingState.toJS();
    }

    return prevRoutingStateJS;
  };
};

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectError,
  makeSelectForecasts,
  makeSelectLocationState,
};
