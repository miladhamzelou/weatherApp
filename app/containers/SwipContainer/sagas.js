/**
 * Gets the forecast data from OpenWeather API
 */

import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_FORECAST } from 'containers/App/constants';
import { forecastLoaded, forecastLoadingError } from 'containers/App/actions';
import request from 'utils/request';
// import { API_KEY } from 'appConfig';

export function* getForecastDataRequest() {
  // direct connection with API does't work with ios/https's missing so we have a proxy
  // const requestURL = `http://api.openweathermap.org/data/2.5/forecast?id=2643743&appid=${API_KEY}&units=metric`;
  const requestURL = 'https://us3r.io:443/proxy';
  try {
    const forecasts = yield call(request, requestURL);
    yield put(forecastLoaded(forecasts));
  } catch (err) {
    yield put(forecastLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* startUpSaga() {
  const watcher = yield takeLatest(LOAD_FORECAST, getForecastDataRequest);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  startUpSaga,
];
