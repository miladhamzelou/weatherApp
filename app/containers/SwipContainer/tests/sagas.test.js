/**
 * Tests for SwipContainer sagas
 */
import { put } from 'redux-saga/effects';
import { forecastLoaded, forecastLoadingError } from 'containers/App/actions';
import { getForecastDataRequest } from '../sagas';

/* eslint-disable redux-saga/yield-effects */
describe('getForecastDataRequest Saga', () => {
  let getForecastsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getForecastsGenerator = getForecastDataRequest();

    const selectDescriptor = getForecastsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the forecastLoaded action if it requests the data successfully', () => {
    const response = {
      list: [{
        dt: 1490724000,
      }],
    };
    const putDescriptor = getForecastsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(forecastLoaded(response)));
  });

  it('should call the forecastLoadingError action if the response errors', () => {
    const response = new Error('An error');
    const putDescriptor = getForecastsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(forecastLoadingError(response)));
  });
});
