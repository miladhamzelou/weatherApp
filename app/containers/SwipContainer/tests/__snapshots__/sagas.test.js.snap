exports[`getForecastDataRequest Saga should call the forecastLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://us3r.io:443/proxy",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getForecastDataRequest Saga should dispatch the forecastLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://us3r.io:443/proxy",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
